runtime/src/main/java/org/corfudb/runtime/object/SnapshotProxy.java

```diff
@@ -3,31 +3,35 @@
 import lombok.NonNull;
 import lombok.extern.slf4j.Slf4j;
 import org.corfudb.protocols.logprotocol.SMREntry;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuError;
 
 import javax.annotation.concurrent.NotThreadSafe;
 import java.util.Map;
 import java.util.function.LongConsumer;
+import java.util.function.Supplier;
 
 @NotThreadSafe
 @Slf4j
-public class SnapshotProxy<T> implements ICorfuSMRSnapshotProxy<T> {
+public class SnapshotProxy<T extends AutoCloseable> implements ICorfuSMRSnapshotProxy<T> {
 
-    private T snapshot;
-
-    private final long baseSnapshotVersion;
+    private final SMRSnapshot<T> snapshot;
+    private T snapshotView;
+    private final Supplier<Long> snapshotVersionSupplier;
 
     private final Map<String, ICorfuSMRUpcallTarget<T>> upcallTargetMap;
 
-    public SnapshotProxy(@NonNull final T snapshot, final long baseSnapshotVersion,
+    public SnapshotProxy(@NonNull final SMRSnapshot<T> snapshot, Supplier<Long> snapshotVersionSupplier,
                          @NonNull final Map<String, ICorfuSMRUpcallTarget<T>> upcallTargetMap) {
         this.snapshot = snapshot;
-        this.baseSnapshotVersion = baseSnapshotVersion;
+        this.snapshotView = snapshot.consume();
+        this.snapshotVersionSupplier = snapshotVersionSupplier;
         this.upcallTargetMap = upcallTargetMap;
     }
 
-    public <R> R access(@NonNull ICorfuSMRAccess<R, T> accessFunction, @NonNull LongConsumer versionAccessed) {
-        final R ret = accessFunction.access(snapshot);
-        versionAccessed.accept(baseSnapshotVersion);
+    public <R> R access(@NonNull ICorfuSMRAccess<R, T> accessFunction,
+                        @NonNull LongConsumer versionAccessed) {
+        final R ret = accessFunction.access(snapshotView);
+        versionAccessed.accept(snapshotVersionSupplier.get());
```

Comment:
This looks like a race condition. Imagine the table is at version=100. T1 performs an access. While this is occurring, T2 performs a sync and updates this table from version=100 to version=150. Then T1 finishes the access and goes to accept the version given from the supplier, which will now be materializedUpTo=150.

I think this can cause missed DCNs in the context of T1 performing a fullSync while at version=100. The call the commit would return 150 in this example, which is where subsequent deltas would be delivered from.